controllers/goroutines/operator_status.go

```diff
@@ -82,37 +83,64 @@ func UpdateCsCrStatus(bs *bootstrap.Bootstrap) {
 	}
 }
 
-func getBedrockOperator(bs *bootstrap.Bootstrap, name, namespace string) apiv3.BedrockOperator {
+func getBedrockOperator(bs *bootstrap.Bootstrap, name, namespace string) (apiv3.BedrockOperator, error) {
 	var opt apiv3.BedrockOperator
+	opt.Name = name
 
 	// fetch subscription
 	sub := &olmv1alpha1.Subscription{}
 	subKey := types.NamespacedName{
 		Name:      name,
 		Namespace: namespace,
 	}
-	if subErr := bs.Client.Get(ctx, subKey, sub); subErr != nil {
-		return opt
+	if err := bs.Client.Get(ctx, subKey, sub); err != nil {
+		return opt, err
 	}
-	installCSV := sub.Status.InstalledCSV
-	if installCSV != "" {
-		opt.Name = installCSV[:strings.IndexByte(installCSV, '.')]
-		opt.Version = installCSV[strings.IndexByte(installCSV, '.')+1:]
+	installedCSV := sub.Status.InstalledCSV
+	if installedCSV != "" {
+		opt.Name = installedCSV[:strings.IndexByte(installedCSV, '.')]
+		opt.Version = installedCSV[strings.IndexByte(installedCSV, '.')+1:]
 	}
 
 	// fetch csv
 	csv := &olmv1alpha1.ClusterServiceVersion{}
 	csvKey := types.NamespacedName{
-		Name:      installCSV,
+		Name:      installedCSV,
 		Namespace: namespace,
 	}
-	if csvErr := bs.Reader.Get(ctx, csvKey, csv); csvErr != nil {
-		return opt
+	if err := bs.Reader.Get(ctx, csvKey, csv); err != nil {
+		klog.Warningf("%s CSV %s", name, err)
+	}
+
+	// fetch installplanName
+	installplanName := ""
+	if sub.Status.Install != nil {
+		installplanName = sub.Status.Install.Name
 	}
+	opt.InstallPlanName = installplanName
+
+	// determinate subscription status
+	if installplanName == "" {
+		opt.SubscriptionStatus = "Failed"
+		opt.InstallPlanName = "Not Found"
+	} else {
+		currentCSV := sub.Status.CurrentCSV
+		installedIsLarger := util.CompareVersion(installedCSV, currentCSV)
+		currentIsLarger := util.CompareVersion(currentCSV, installedCSV)
+		if !installedIsLarger && !currentIsLarger {
+			// installedCSV == currentCSV
+			opt.SubscriptionStatus = "Succeeded"
+		} else if currentIsLarger {
+			opt.SubscriptionStatus = "Upgrade available"
+		} else {
+			klog.Warning("installedCSV version is larger than currentCSV")
+		}
+	}
+
 	if len(csv.Status.Conditions) > 0 {
```

Comment:
```
if err := bs.Reader.Get(ctx, csvKey, csv); err != nil {
	klog.Warningf("%s CSV %s", name, err)
}
```

Since we did not return error above when we try to get csv, I am wondering whether we will hit `nil` pointer issue when we check `len(csv.Status.Conditions)`?

I think maybe we could leave `opt.Status` blank if we failed to get CSV instead of directly return the error?

Reply:
csv is pointer in `csv.Status.Conditions`. We will hit nil pointer panic.

Reply:
good catch, I will update the checking

Reply:
Sorry, I was wrong, csv will never be nil, since it's initialized in `csv := &olmv1alpha1.ClusterServiceVersion{}`.
Thus we won't get nil pointer issue, like https://play.golang.org/p/D876UJFcoYi
But it's better to make it more clear in the code.